apiserver/application/application_unit_test.go

```diff
@@ -252,11 +316,175 @@ func (s *ApplicationSuite) TestDestroyUnit(c *gc.C) {
 	}})
 }
 
+func (s *ApplicationSuite) setUpSeriesSubordinates() {
+	ch := mockCharm{
+		meta: &charm.Meta{
+			Series: []string{"trusty", "xenial"},
+		},
+	}
+	app := mockApplication{
+		charm:       &ch,
+		charmURL:    "cs:foo-sub",
+		subordinate: true,
+		units: []mockUnit{{
+			tag: names.NewUnitTag("foo-sub/0"),
+		}, {
+			tag: names.NewUnitTag("foo-sub/1"),
+		}},
+	}
+	s.charms["cs:foo-sub"] = &ch
+	s.applications["foo-sub"] = &app
+	s.backend.charms["cs:foo-sub"] = &ch
+	s.backend.applications["foo-sub"] = &app
+	for i, u := range s.applications["foo"].units {
+		u.subordinateNames = []string{fmt.Sprintf("foo-sub/%d", i)}
+	}
+}
+
+func (s *ApplicationSuite) TestApplicationUpdateSeries(c *gc.C) {
+	s.setUpBasicTest()
+	args := params.ApplicationUpdateSeriesArgs{
+		Args: []params.ApplicationUpdateSeriesArg{{
+			ApplicationName: "foo",
+			Series:          "trusty",
+		}, {
+			ApplicationName: "foo",
+			Series:          "zesty",
+		}},
+	}
+	results, err := s.api.UpdateApplicationSeries(args)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(len(results.Results), gc.Equals, 2)
+	c.Assert(results.Results[0], jc.DeepEquals, params.ErrorResult{})
+	c.Assert(results.Results[1], jc.DeepEquals, params.ErrorResult{
+		Error: &params.Error{
+			Code:    params.CodeIncompatibleSeries,
+			Message: `"zesty" is incompatible`,
+		},
+	})
+	s.backend.CheckCallNames(c, "ModelTag", "Application", "Application")
+	app := s.applications["foo"]
+	app.CheckCallNames(c, "IsPrincipal", "VerifySupportedSeries", "AllUnits", "UpdateApplicationSeries", "IsPrincipal", "VerifySupportedSeries")
+}
+
+func (s *ApplicationSuite) TestApplicationUpdateSeriesWithSubordinates(c *gc.C) {
+	s.setUpBasicTest()
+	s.setUpSeriesSubordinates()
+	args := params.ApplicationUpdateSeriesArgs{
+		Args: []params.ApplicationUpdateSeriesArg{{
+			ApplicationName: "foo",
+			Series:          "xenial",
+		}},
+	}
+	results, err := s.api.UpdateApplicationSeries(args)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(len(results.Results), gc.Equals, 1)
+	c.Assert(results.Results[0], jc.DeepEquals, params.ErrorResult{})
+	s.backend.CheckCallNames(c, "ModelTag", "Application")
+	app := s.applications["foo"]
+	app.CheckCallNames(c, "IsPrincipal", "VerifySupportedSeries", "AllUnits", "UpdateApplicationSeries")
+	unit := app.units[0]
+	unit.CheckCallNames(c, "SubordinateNames")
+}
+
+func (s *ApplicationSuite) TestApplicationUpdateSeriesWithSubordinatesFail(c *gc.C) {
+	s.setUpBasicTest()
+	s.setUpSeriesSubordinates()
+	args := params.ApplicationUpdateSeriesArgs{
+		Args: []params.ApplicationUpdateSeriesArg{{
+			ApplicationName: "foo",
+			Series:          "yakkety",
+		}},
+	}
+	results, err := s.api.UpdateApplicationSeries(args)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(len(results.Results), gc.Equals, 1)
+	c.Assert(results.Results[0], jc.DeepEquals, params.ErrorResult{})
```

Comment:
err is used to indicate a problem found within the apiserver call, while results (param.ErrorResults)  is passed as an argument with the facade call to capture the individual results per application when the related function in state is called.  This pattern has been used by the other bulk API calls.